File: apps/research-ui/decoder_service.py
import base64
import binascii
import json
import os
import struct
import subprocess
import threading
import time
from pathlib import Path
from typing import Any, Optional

DECODER_PROTOCOL_VERSION = 1
MAX_DECODER_INPUT_BYTES = 1 << 20
MAX_DECODER_OUTPUT_BYTES = 1 << 20
MAX_DECODER_ACTION_BYTES = 1536 << 10
MAX_DECODER_PIPELINE_STEPS = 16
MAX_DECODER_RETAINED_BYTES = 4 << 20
MAX_DECODER_JWT_BYTES = 64 << 10
MAX_DECODER_SECRET_BYTES = 4 << 10
MAX_DECODER_JSON_DEPTH = 64
MAX_DECODER_JSON_TOKENS = 100_000
DECODER_TIMEOUT_SECONDS = 2.0
MAX_JWT_EXPIRY_SECONDS = 7 * 24 * 60 * 60

DECODER_OPERATIONS = (
    "base64-encode",
    "base64-decode",
    "base64url-encode",
    "base64url-decode",
    "hex-encode",
    "hex-decode",
    "url-encode",
    "url-decode",
    "base36-encode",
    "base36-decode",
    "gzip-compress",
    "gzip-decompress",
    "zlib-compress",
    "zlib-decompress",
    "deflate-compress",
    "deflate-decompress",
    "json-pretty",
    "json-minify",
)
JWT_ALGORITHMS = ("HS256", "HS384", "HS512", "none")


class DecoderError(ValueError):
    pass


class DecoderProtocolError(DecoderError):
    pass


class DecoderUnavailable(DecoderError):
    pass


class DecoderTimeout(DecoderError):
    pass


def decoder_limits() -> dict[str, Any]:
    return {
        "input_bytes": MAX_DECODER_INPUT_BYTES,
        "output_bytes": MAX_DECODER_OUTPUT_BYTES,
        "pipeline_steps": MAX_DECODER_PIPELINE_STEPS,
        "retained_bytes": MAX_DECODER_RETAINED_BYTES,
        "jwt_bytes": MAX_DECODER_JWT_BYTES,
        "secret_bytes": MAX_DECODER_SECRET_BYTES,
        "json_depth": MAX_DECODER_JSON_DEPTH,
        "json_tokens": MAX_DECODER_JSON_TOKENS,
        "timeout_ms": int(DECODER_TIMEOUT_SECONDS * 1_000),
        "operations": list(DECODER_OPERATIONS),
        "jwt_algorithms": list(JWT_ALGORITHMS),
    }


def _bounded_text(value: Any, label: str, maximum_bytes: int) -> str:
    if not isinstance(value, str):
        raise DecoderError(f"{label} must be text")
    try:
        encoded = value.encode("utf-8")
    except UnicodeEncodeError as exception:
        raise DecoderError(f"{label} must be valid UTF-8") from exception
    if len(encoded) > maximum_bytes:
        raise DecoderError(f"{label} exceeds {maximum_bytes} bytes")
    return value


def _decode_base64(value: Any, label: str, maximum_bytes: int) -> bytes:
    text = _bounded_text(value, label, ((maximum_bytes + 2) // 3) * 4 + 4)
    try:
        decoded = base64.b64decode(text, validate=True)
    except (binascii.Error, ValueError) as exception:
        raise DecoderError(f"{label} must be canonical Base64") from exception
    if len(decoded) > maximum_bytes:
        raise DecoderError(f"{label} exceeds {maximum_bytes} decoded bytes")
    return decoded


def _frame(*parts: bytes) -> bytes:
    return b"".join(struct.pack(">I", len(part)) + part for part in parts)


def _utf8_or_none(value: bytes) -> Optional[str]:
    try:
        return value.decode("utf-8")
    except UnicodeDecodeError:
        return None


class DecoderService:
    def __init__(self, executable: Path):
        self.executable = executable.resolve()
        self._lock = threading.Lock()

    def available(self) -> bool:
        return self.executable.is_file() and os.access(self.executable, os.X_OK)

    def state(self) -> dict[str, Any]:
        return {
            "protocol_version": DECODER_PROTOCOL_VERSION,
            "available": self.available(),
            "busy": self._lock.locked(),
            "limits": decoder_limits(),
        }

    def action(self, request: dict[str, Any]) -> dict[str, Any]:
        if not isinstance(request, dict):
            raise DecoderError("Decoder action must be an object")
        if request.get("protocol_version") != DECODER_PROTOCOL_VERSION:
            raise DecoderError("Decoder protocol version is unsupported")
        action = request.get("action")
        if action == "transform":
            return self._transform(request)
        if action == "jwt_inspect":
            return self._jwt_inspect(request)
        if action == "jwt_verify":
            return self._jwt_verify(request)
        if action == "jwt_create":
            return self._jwt_create(request)
        raise DecoderError("Decoder action is unsupported")

    def _run(self, arguments: list[str], input_bytes: bytes) -> subprocess.CompletedProcess:
        if not self.available():
            raise DecoderUnavailable("The native decoder executable is unavailable")
        started = time.monotonic_ns()
        acquired = self._lock.acquire(timeout=DECODER_TIMEOUT_SECONDS)
        if not acquired:
            raise DecoderTimeout("The native decoder is busy")
        try:
            try:
                process = subprocess.run(
                    [str(self.executable), *arguments],
                    input=input_bytes,
                    capture_output=True,
                    check=False,
                    timeout=DECODER_TIMEOUT_SECONDS,
                    env={"LC_ALL": "C", "PATH": os.defpath},
                )
            except subprocess.TimeoutExpired as exception:
                raise DecoderTimeout("Decoder operation exceeded two seconds") from exception
        finally:
            self._lock.release()
        process.duration_us = max(1, (time.monotonic_ns() - started) // 1_000)
        return process

    def _transform(self, request: dict[str, Any]) -> dict[str, Any]:
        if set(request) != {
            "protocol_version",
            "action",
            "operation_id",
            "operation",
            "input_base64",
        }:
            raise DecoderError("Decoder transform shape is invalid")
        operation_id = request.get("operation_id")
        if not isinstance(operation_id, int) or isinstance(operation_id, bool) or operation_id < 1:
            raise DecoderError("Decoder operation ID must be a positive integer")
        operation = request.get("operation")
        if operation not in DECODER_OPERATIONS:
            raise DecoderError("Decoder transform is not allowlisted")
        input_bytes = _decode_base64(
            request.get("input_base64"), "Decoder input", MAX_DECODER_INPUT_BYTES
        )
        process = self._run(["transform", operation], input_bytes)
        if process.returncode != 0:
            error = process.stderr[:4_096].decode("utf-8", errors="replace").strip()
            if process.returncode == 3:
                raise DecoderProtocolError(error or "Decoder output limit was reached")
            raise DecoderError(error or "Native decoder rejected the transform")
        if len(process.stdout) > MAX_DECODER_OUTPUT_BYTES:
            raise DecoderProtocolError("Native decoder returned oversized output")
        text = _utf8_or_none(process.stdout)
        return {
            "protocol_version": DECODER_PROTOCOL_VERSION,
            "ok": True,
            "operation_id": operation_id,
            "operation": operation,
            "input_bytes": len(input_bytes),
            "output_bytes": len(process.stdout),
            "output_base64": base64.b64encode(process.stdout).decode("ascii"),
            "utf8_text": text,
            "hex_preview": process.stdout[:256].hex(),
            "preview_truncated": len(process.stdout) > 256,
            "duration_us": process.duration_us,
        }

    def _jwt_inspect(self, request: dict[str, Any]) -> dict[str, Any]:
        if set(request) != {"protocol_version", "action", "token"}:
            raise DecoderError("JWT inspection shape is invalid")
        token = _bounded_text(request.get("token"), "JWT", MAX_DECODER_JWT_BYTES)
        return self._jwt_response(self._run(["jwt-inspect"], token.encode("utf-8")))

    def _jwt_verify(self, request: dict[str, Any]) -> dict[str, Any]:
        if set(request) != {"protocol_version", "action", "token", "secret"}:
            raise DecoderError("JWT verification shape is invalid")
        token = _bounded_text(request.get("token"), "JWT", MAX_DECODER_JWT_BYTES)
        secret = _bounded_text(
            request.get("secret"), "JWT HMAC secret", MAX_DECODER_SECRET_BYTES
        ).encode("utf-8")
        if not secret:
            raise DecoderError("JWT HMAC secret cannot be empty")
        return self._jwt_response(
            self._run(["jwt-verify"], _frame(token.encode("utf-8"), secret))
        )

    def _jwt_create(self, request: dict[str, Any]) -> dict[str, Any]:
        if set(request) != {
            "protocol_version",
            "action",
            "payload_json",
            "algorithm",
            "secret",
            "expires_in_seconds",
            "allow_unsigned_confirmed",
        }:
            raise DecoderError("JWT creation shape is invalid")
        payload = _bounded_text(
            request.get("payload_json"), "JWT payload", 56 << 10
        ).encode("utf-8")
        algorithm = request.get("algorithm")
        if algorithm not in JWT_ALGORITHMS:
            raise DecoderError("JWT creation algorithm is unsupported")
        secret = _bounded_text(
            request.get("secret"), "JWT HMAC secret", MAX_DECODER_SECRET_BYTES
        ).encode("utf-8")
        unsigned_confirmed = request.get("allow_unsigned_confirmed")
        if not isinstance(unsigned_confirmed, bool):
            raise DecoderError("Unsigned JWT confirmation must be boolean")
        if algorithm == "none":
            if not unsigned_confirmed:
                raise DecoderError("Creating an unsigned JWT requires explicit confirmation")
            if secret:
                raise DecoderError("Unsigned JWT creation does not accept a secret")
        elif not secret:
            raise DecoderError("Signed JWT creation requires an HMAC secret")
        expires_in = request.get("expires_in_seconds")
        expiration = "none"
        if expires_in is not None:
            if (
                not isinstance(expires_in, int)
                or isinstance(expires_in, bool)
                or not 1 <= expires_in <= MAX_JWT_EXPIRY_SECONDS
            ):
                raise DecoderError("JWT expiry must be between one second and seven days")
            expiration = str(int(time.time()) + expires_in)
        return self._jwt_response(
            self._run(
                ["jwt-create", algorithm, expiration],
                _frame(payload, secret),
            ),
            creation=True,
        )

    def _jwt_response(
        self, process: subprocess.CompletedProcess, creation: bool = False
    ) -> dict[str, Any]:
        if process.returncode != 0 or len(process.stdout) > 256 << 10:
            raise DecoderProtocolError("Native JWT decoder returned an invalid response")
        try:
            response = json.loads(process.stdout)
        except (UnicodeDecodeError, json.JSONDecodeError) as exception:
            raise DecoderProtocolError("Native JWT decoder returned malformed JSON") from exception
        expected = (
            {"protocol_version", "ok", "token", "error"}
            if creation
            else {
                "protocol_version",
                "ok",
                "algorithm",
                "signature_status",
                "header_json",
                "payload_json",
                "token_bytes",
                "signature_bytes",
                "error",
            }
        )
        if not isinstance(response, dict) or set(response) != expected:
            raise DecoderProtocolError("Native JWT decoder response contract is invalid")
        response["duration_us"] = process.duration_us
        return response

File: apps/research-ui/macos/DecoderService.swift
import CoreFoundation
import Darwin
import Foundation

struct NativeDecoderError: Error {
  let status: Int
  let message: String
}

final class NativeDecoderService {
  private static let inputBytes = 1 << 20
  private static let outputBytes = 1 << 20
  private static let jwtBytes = 64 << 10
  private static let secretBytes = 4 << 10
  private static let timeoutSeconds = 2.0
  private static let operations = [
    "base64-encode", "base64-decode", "base64url-encode", "base64url-decode",
    "hex-encode", "hex-decode", "url-encode", "url-decode", "base36-encode",
    "base36-decode", "gzip-compress", "gzip-decompress", "zlib-compress",
    "zlib-decompress", "deflate-compress", "deflate-decompress", "json-pretty",
    "json-minify",
  ]
  private static let algorithms = ["HS256", "HS384", "HS512", "none"]

  private let executableURL: URL
  private let lock = NSLock()

  init(executableURL: URL) {
    self.executableURL = executableURL
  }

  func state() throws -> Data {
    let idle = lock.try()
    if idle { lock.unlock() }
    return try JSONSerialization.data(
      withJSONObject: [
        "protocol_version": 1,
        "available": available,
        "busy": !idle,
        "limits": Self.limits,
      ]
    )
  }

  var available: Bool {
    FileManager.default.isExecutableFile(atPath: executableURL.path)
  }

  func action(_ request: [String: Any]) throws -> Data {
    guard integer(request["protocol_version"]) == 1, let action = request["action"] as? String else {
      throw NativeDecoderError(status: 400, message: "Decoder protocol version is unsupported")
    }
    switch action {
      case "transform":
        return try transform(request)
      case "jwt_inspect":
        return try inspectJWT(request)
      case "jwt_verify":
        return try verifyJWT(request)
      case "jwt_create":
        return try createJWT(request)
      default:
        throw NativeDecoderError(status: 400, message: "Decoder action is unsupported")
    }
  }

  private static var limits: [String: Any] {
    [
      "input_bytes": inputBytes,
      "output_bytes": outputBytes,
      "pipeline_steps": 16,
      "retained_bytes": 4 << 20,
      "jwt_bytes": jwtBytes,
      "secret_bytes": secretBytes,
      "json_depth": 64,
      "json_tokens": 100_000,
      "timeout_ms": 2_000,
      "operations": operations,
      "jwt_algorithms": algorithms,
    ]
  }

  private func transform(_ request: [String: Any]) throws -> Data {
    guard exactKeys(
      request,
      ["protocol_version", "action", "operation_id", "operation", "input_base64"]
    ), let operationID = integer(request["operation_id"]), operationID >= 1,
      let operation = request["operation"] as? String, Self.operations.contains(operation),
      let input = canonicalBase64(request["input_base64"], maximumBytes: Self.inputBytes)
    else {
      throw NativeDecoderError(status: 400, message: "Decoder transform shape is invalid")
    }
    let result = try run(arguments: ["transform", operation], input: input)
    guard result.status == 0 else {
      let error = boundedUTF8(result.error, maximumBytes: 4_096)
      throw NativeDecoderError(
        status: result.status == 3 ? 422 : 400,
        message: error.isEmpty ? "Native decoder rejected the transform" : error
      )
    }
    guard result.output.count <= Self.outputBytes else {
      throw NativeDecoderError(status: 422, message: "Native decoder returned oversized output")
    }
    let text = String(data: result.output, encoding: .utf8)
    return try JSONSerialization.data(
      withJSONObject: [
        "protocol_version": 1,
        "ok": true,
        "operation_id": operationID,
        "operation": operation,
        "input_bytes": input.count,
        "output_bytes": result.output.count,
        "output_base64": result.output.base64EncodedString(),
        "utf8_text": text ?? NSNull(),
        "hex_preview": result.output.prefix(256).map { String(format: "%02x", $0) }.joined(),
        "preview_truncated": result.output.count > 256,
        "duration_us": result.durationMicroseconds,
      ]
    )
  }

  private func inspectJWT(_ request: [String: Any]) throws -> Data {
    guard exactKeys(request, ["protocol_version", "action", "token"]),
      let token = boundedText(request["token"], maximumBytes: Self.jwtBytes)
    else {
      throw NativeDecoderError(status: 400, message: "JWT inspection shape is invalid")
    }
    return try jwtResponse(
      run(arguments: ["jwt-inspect"], input: Data(token.utf8)),
      creation: false
    )
  }

  private func verifyJWT(_ request: [String: Any]) throws -> Data {
    guard exactKeys(request, ["protocol_version", "action", "token", "secret"]),
      let token = boundedText(request["token"], maximumBytes: Self.jwtBytes),
      let secret = boundedText(request["secret"], maximumBytes: Self.secretBytes),
      !secret.isEmpty
    else {
      throw NativeDecoderError(status: 400, message: "JWT verification shape is invalid")
    }
    return try jwtResponse(
      run(
        arguments: ["jwt-verify"],
        input: frame([Data(token.utf8), Data(secret.utf8)])
      ),
      creation: false
    )
  }

  private func createJWT(_ request: [String: Any]) throws -> Data {
    guard exactKeys(
      request,
      [
        "protocol_version", "action", "payload_json", "algorithm", "secret",
        "expires_in_seconds", "allow_unsigned_confirmed",
      ]
    ), let payload = boundedText(request["payload_json"], maximumBytes: 56 << 10),
      let algorithm = request["algorithm"] as? String, Self.algorithms.contains(algorithm),
      let secret = boundedText(request["secret"], maximumBytes: Self.secretBytes),
      let unsignedConfirmed = request["allow_unsigned_confirmed"] as? Bool
    else {
      throw NativeDecoderError(status: 400, message: "JWT creation shape is invalid")
    }
    if algorithm == "none" {
      guard unsignedConfirmed, secret.isEmpty else {
        throw NativeDecoderError(
          status: 400,
          message: "Creating an unsigned JWT requires confirmation and no secret"
        )
      }
    } else if secret.isEmpty {
      throw NativeDecoderError(status: 400, message: "Signed JWT creation requires a secret")
    }
    let expiration: String
    if request["expires_in_seconds"] is NSNull {
      expiration = "none"
    } else {
      guard let seconds = integer(request["expires_in_seconds"]), (1...604_800).contains(seconds) else {
        throw NativeDecoderError(
          status: 400,
          message: "JWT expiry must be between one second and seven days"
        )
      }
      expiration = String(Int(Date().timeIntervalSince1970) + seconds)
    }
    return try jwtResponse(
      run(
        arguments: ["jwt-create", algorithm, expiration],
        input: frame([Data(payload.utf8), Data(secret.utf8)])
      ),
      creation: true
    )
  }

  private func jwtResponse(_ result: ProcessResult, creation: Bool) throws -> Data {
    guard result.status == 0, result.output.count <= 256 << 10,
      var response = try JSONSerialization.jsonObject(with: result.output) as? [String: Any]
    else {
      throw NativeDecoderError(status: 422, message: "Native JWT decoder returned an invalid response")
    }
    let expected: Set<String> = creation
      ? ["protocol_version", "ok", "token", "error"]
      : [
          "protocol_version", "ok", "algorithm", "signature_status", "header_json",
          "payload_json", "token_bytes", "signature_bytes", "error",
        ]
    guard Set(response.keys) == expected, integer(response["protocol_version"]) == 1,
      response["ok"] is Bool
    else {
      throw NativeDecoderError(status: 422, message: "Native JWT decoder contract is invalid")
    }
    response["duration_us"] = result.durationMicroseconds
    return try JSONSerialization.data(withJSONObject: response)
  }

  private struct ProcessResult {
    let output: Data
    let error: Data
    let status: Int32
    let durationMicroseconds: Int
  }

  private func run(arguments: [String], input: Data) throws -> ProcessResult {
    guard available else {
      throw NativeDecoderError(status: 503, message: "The packaged native decoder is unavailable")
    }
    guard lock.lock(before: Date().addingTimeInterval(Self.timeoutSeconds)) else {
      throw NativeDecoderError(status: 408, message: "The packaged native decoder is busy")
    }
    defer { lock.unlock() }
    let process = Process()
    process.executableURL = executableURL
    process.arguments = arguments
    process.environment = ["LANG": "C", "LC_ALL": "C", "TZ": "UTC"]
    let inputPipe = Pipe()
    let outputPipe = Pipe()
    let errorPipe = Pipe()
    process.standardInput = inputPipe
    process.standardOutput = outputPipe
    process.standardError = errorPipe
    let terminated = DispatchSemaphore(value: 0)
    process.terminationHandler = { _ in terminated.signal() }
    do {
      try process.run()
    } catch {
      throw NativeDecoderError(status: 500, message: "Could not start the packaged native decoder")
    }
    let outputGroup = DispatchGroup()
    let outputLock = NSLock()
    var output = Data()
    var error = Data()
    outputGroup.enter()
    DispatchQueue.global(qos: .userInitiated).async {
      let data = outputPipe.fileHandleForReading.readDataToEndOfFile()
      outputLock.lock()
      output = data
      outputLock.unlock()
      outputGroup.leave()
    }
    outputGroup.enter()
    DispatchQueue.global(qos: .utility).async {
      let data = errorPipe.fileHandleForReading.readDataToEndOfFile()
      outputLock.lock()
      error = data
      outputLock.unlock()
      outputGroup.leave()
    }
    let started = DispatchTime.now().uptimeNanoseconds
    inputPipe.fileHandleForWriting.write(input)
    try? inputPipe.fileHandleForWriting.close()
    let timedOut = terminated.wait(timeout: .now() + Self.timeoutSeconds) == .timedOut
    if timedOut {
      process.terminate()
      if terminated.wait(timeout: .now() + 0.25) == .timedOut {
        kill(process.processIdentifier, SIGKILL)
        _ = terminated.wait(timeout: .now() + 0.25)
      }
    }
    outputGroup.wait()
    if timedOut {
      throw NativeDecoderError(status: 408, message: "Decoder operation exceeded two seconds")
    }
    let elapsed = DispatchTime.now().uptimeNanoseconds - started
    return ProcessResult(
      output: output,
      error: error,
      status: process.terminationStatus,
      durationMicroseconds: max(1, Int(elapsed / 1_000))
    )
  }

  private func canonicalBase64(_ value: Any?, maximumBytes: Int) -> Data? {
    guard let text = boundedText(value, maximumBytes: ((maximumBytes + 2) / 3) * 4 + 4),
      let data = Data(base64Encoded: text), data.count <= maximumBytes,
      data.base64EncodedString() == text
    else { return nil }
    return data
  }

  private func frame(_ values: [Data]) -> Data {
    var result = Data()
    for value in values {
      var size = UInt32(value.count).bigEndian
      withUnsafeBytes(of: &size) { result.append(contentsOf: $0) }
      result.append(value)
    }
    return result
  }

  private func exactKeys(_ value: [String: Any], _ keys: Set<String>) -> Bool {
    Set(value.keys) == keys
  }

  private func integer(_ value: Any?) -> Int? {
    guard let number = value as? NSNumber,
      CFGetTypeID(number) != CFBooleanGetTypeID()
    else { return nil }
    let double = number.doubleValue
    guard double.isFinite, double.rounded() == double,
      double >= Double(Int.min), double <= Double(Int.max)
    else { return nil }
    return number.intValue
  }

  private func boundedText(_ value: Any?, maximumBytes: Int) -> String? {
    guard let text = value as? String, text.utf8.count <= maximumBytes else { return nil }
    return text
  }

  private func boundedUTF8(_ value: Data, maximumBytes: Int) -> String {
    String(data: Data(value.prefix(maximumBytes)), encoding: .utf8)?
      .trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
  }
}
